src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp

```diff
@@ -925,29 +945,38 @@ RangesInDataParts MergeTreeDataSelectExecutor::filterPartsByPrimaryKeyAndSkipInd
         });
     }
 
-    for (size_t idx = 0; idx < skip_indexes.useful_indices.size(); ++idx)
+    const auto num_indices = skip_indexes.useful_indices.size();
+
+    const auto part_stats_granularity = settings[Setting::per_part_index_stats] ? original_num_parts : 1;
+    for (size_t part_index = 0; part_index < part_stats_granularity; ++part_index)
     {
-        const auto & index_and_condition = skip_indexes.useful_indices[idx];
-        const auto & stat = useful_indices_stat[idx];
-        const auto & index_name = index_and_condition.index->index.name;
-        LOG_DEBUG(
-            log,
-            "Index {} has dropped {}/{} granules, it took {}ms across {} threads.",
-            backQuote(index_name),
-            stat.granules_dropped.load(),
-            stat.total_granules.load(),
-            stat.elapsed_us.load() / 1000,
-            num_threads);
+        for (size_t idx = 0; idx < skip_indexes.useful_indices.size(); ++idx)
+        {
+            const auto & stat = useful_indices_stat[part_index * num_indices + idx];
+            const auto & index_and_condition = skip_indexes.useful_indices[skip_indexes.per_part_index_orders[part_index][idx]];
+            const auto & index_name = index_and_condition.index->index.name;
+            LOG_DEBUG(
+                log,
+                "Index {} has dropped {}/{} granules, it took {}ms across {} threads.",
+                backQuote(index_name),
+                stat.granules_dropped.load(),
+                stat.total_granules.load(),
+                stat.elapsed_us.load() / 1000,
+                num_threads);
 
-        std::string description
-            = index_and_condition.index->index.type + " GRANULARITY " + std::to_string(index_and_condition.index->index.granularity);
+            std::string description = index_and_condition.index->index.type + " GRANULARITY " + std::to_string(index_and_condition.index->index.granularity);
+            if (settings[Setting::per_part_index_stats])
+            {
+                description += " PART " + std::to_string(part_index) + "/" + std::to_string(part_stats_granularity);
+            }
 
-        index_stats.emplace_back(ReadFromMergeTree::IndexStat{
-            .type = ReadFromMergeTree::IndexType::Skip,
-            .name = index_name,
-            .description = std::move(description),
-            .num_parts_after = stat.total_parts - stat.parts_dropped,
-            .num_granules_after = stat.total_granules - stat.granules_dropped});
+            index_stats.emplace_back(ReadFromMergeTree::IndexStat{
```

Comment:
I think in case of `per_part_index_stats = 1`, it makes sense to propagate the part name to the stats.